app/controllers/claimToAdjustPoa/AmendablePOAController.scala

```diff
@@ -49,26 +50,44 @@ class AmendablePOAController @Inject()(val authorisedFunctions: AuthorisedFuncti
   def show(isAgent: Boolean): Action[AnyContent] =
     auth.authenticatedAction(isAgent) {
       implicit user =>
-        ifAdjustPoaIsEnabled(isAgent) {
-          claimToAdjustService.getPoaForNonCrystallisedTaxYear(Nino(user.nino)) flatMap {
-            case Right(Some(paymentOnAccount: PaymentOnAccountViewModel)) =>
-              Future.successful(
-                Ok(view(
-                  isAgent = isAgent,
-                  paymentOnAccount = paymentOnAccount
-                ))
-              )
-            case Right(None) =>
-              Logger("application").error(s"Failed to create PaymentOnAccount model")
-              Future.successful(showInternalServerError(isAgent))
-            case Left(ex) =>
-              Logger("application").error(s"Exception: ${ex.getMessage} - ${ex.getCause}")
-              Future.failed(ex)
-          }
+        ifAdjustPoaIsEnabled(isAgent) {{
+          for {
+            poaMaybe <- EitherT(claimToAdjustService.getPoaForNonCrystallisedTaxYear(Nino(user.nino)))
+            _ <- EitherT(handleSession)
+          } yield poaMaybe
+        }.value.flatMap {
+          case Right(Some(paymentOnAccount: PaymentOnAccountViewModel)) =>
+            Future.successful(
+              Ok(view(
+                isAgent = isAgent,
+                paymentOnAccount = paymentOnAccount
+              ))
+            )
+          case Right(None) =>
+            Logger("application").error(s"Failed to create PaymentOnAccount model")
+            Future.successful(showInternalServerError(isAgent))
+          case Left(ex) =>
+            Logger("application").error(s"Exception: ${ex.getMessage} - ${ex.getCause}")
+            Future.failed(ex)
+        }
         } recover {
           case ex: Exception =>
             Logger("application").error(s"Unexpected error: ${ex.getMessage} - ${ex.getCause}")
             showInternalServerError(isAgent)
         }
     }
+
+  private def handleSession(implicit hc: HeaderCarrier): Future[Either[Throwable, Unit]] = {
```

Comment:
:thinking: - this method is a good candidate to be moved in the separate trait for re-use in other page in ClaimToAdjust the flow;
:wrench: - please add logging, especially around error cases;